lib/webmock/http_lib_adapters/net_http.rb

```diff
@@ -72,78 +72,57 @@ def constants(inherit=true)
+
           request_signature = WebMock::NetHTTPUtility.request_signature_from_request(self, request, body)
 
           WebMock::RequestRegistry.instance.requested_signatures.put(request_signature)
 
           if webmock_response = WebMock::StubRegistry.instance.response_for_request(request_signature)
-            @socket = Net::HTTP.socket_type.new
             WebMock::CallbackRegistry.invoke_callbacks(
               {lib: :net_http}, request_signature, webmock_response)
             build_net_http_response(webmock_response, &block)
           elsif WebMock.net_connect_allowed?(request_signature.uri)
             check_right_http_connection
-            after_request = lambda do |response|
-              if WebMock::CallbackRegistry.any_callbacks?
-                webmock_response = build_webmock_response(response)
-                WebMock::CallbackRegistry.invoke_callbacks(
-                  {lib: :net_http, real_request: true}, request_signature, webmock_response)
-              end
-              response.extend Net::WebMockHTTPResponse
-              block.call response if block
-              response
-            end
-            super_with_after_request = lambda {
-              response = super(request, nil, &nil)
-              after_request.call(response)
-            }
-            if started?
-              ensure_actual_connection
-              super_with_after_request.call
-            else
-              start_with_connect {
-                super_with_after_request.call
-              }
+            ensure_actually_connected
+
+            response = super(request, nil, &nil)
+
+            if WebMock::CallbackRegistry.any_callbacks?
+              WebMock::CallbackRegistry.invoke_callbacks(
+                {lib: :net_http, real_request: true},
+                request_signature,
+                build_webmock_response(response)
+              )
             end
+
+            response.extend Net::WebMockHTTPResponse
+            block.call response if block
+            response
           else
             raise WebMock::NetConnectNotAllowedError.new(request_signature)
           end
         end
 
-        def start_without_connect
-          raise IOError, 'HTTP session already opened' if @started
-          if block_given?
-            begin
-              @socket = Net::HTTP.socket_type.new
-              @started = true
-              return yield(self)
-            ensure
-              do_finish
-            end
-          end
-          @socket = Net::HTTP.socket_type.new
-          @started = true
-          self
-        end
+        private
 
+        alias_method :actually_connect, :connect
 
-        def ensure_actual_connection
-          if @socket.is_a?(StubSocket)
-            @socket&.close
-            @socket = nil
-            do_start
-          end
-        end
-
-        alias_method :start_with_connect, :start
-
-        def start(&block)
+        def connect
```

Comment:
`Net::HTTP#start` calls `#connect`. This is a much simpler place for us to hook in, because unlike `#start`, `#connect` doesn't have a block variant.

Reply:
👍